File: Project_02_C#/project_2_CSharp/project_2_CSharp/product/Device.cs
using project_2_CSharp.App;
using System;
using System.Collections.Generic;
using System.Text;

namespace project_2_CSharp
{
    abstract class Device
    {
        private String brand;
        private string name;
        private String size;
        private String chip;
        private int ram;
        private int capacity;
        private String color;
        private decimal price;
        private string os;
        public string Brand { get => brand; set => brand = value; }
        public string Size { get => size; set => size = value; }
        public string Chip { get => chip; set => chip = value; }
        public int Ram { get => ram; set => ram = value; }
        public int Capacity { get => capacity; set => capacity = value; }
        public string Color { get => color; set => color = value; }
        public decimal Price { get => price; set => price = value; }
        public string Name { get => name; set => name = value; }
        public string Os { get => os; set => os = value; }

        public Device() { }

        protected Device(string brand, string name, string os, string size, string chip, int ram, int capacity, string color, decimal price)
        {
            Brand = brand;
            Size = size;
            Chip = chip;
            Ram = ram;
            Capacity = capacity;
            Color = color;
            Price = price;
            Name = name;
            Os = os;
        }
        protected Device(string brand, string name, string size, string chip, int ram, int capacity, string color, decimal price)
        {
            Brand = brand;
            Size = size;
            Chip = chip;
            Ram = ram;
            Capacity = capacity;
            Color = color;
            Price = price;
            Name = name;
        }
        public abstract void insert();
        ~Device()
        {
            System.Diagnostics.Trace.WriteLine("Deletted device!!!");
        }
        public abstract void Info();

        public static bool operator <(Device phone1, Device phone2)
        {
            int point = 0;
            if (phone1.Ram < phone2.Ram)
                point++;
            if (phone1.Capacity< phone2.Capacity)
                point++;
            if (phone1.Price > phone2.Price)
                point++;
            if (point >= 2)
                return true;  
            return false;
        }
        public static bool operator >(Device phone1, Device phone2)
        {
            int point = 0;
            if (phone1.Ram > phone2.Ram)
                point++;
            if (phone1.Capacity > phone2.Capacity)
                point++;
            if (phone1.Price < phone2.Price)
                point++;
            if (point >= 2)
                return true;
            return false;
        }
        public virtual string installApp(IAndroidCompatible app) { return ""; }
        public virtual string installApp(IAppleCompatible app) { return ""; }
    }
    }

File: project_2_java/src/product/Device.java
package product;

import app.IAndroidCompatible;
import app.IAppleCompatible;

public abstract class Device {
	private String brand;
	private String name;
	private String size;
	private String chip;
	private int ram;
	protected int capacity;
	private String color;
	private double price;
	private String os;
	public final String getBrand()
	{
		return brand;
	}
	public final void setBrand(String value)
	{
		brand = value;
	}
	public final String getSize()
	{
		return size;
	}
	public final void setSize(String value)
	{
		size = value;
	}
	public final String getChip()
	{
		return chip;
	}
	public final void setChip(String value)
	{
		chip = value;
	}
	public final int getRam()
	{
		return ram;
	}
	public final void setRam(int value)
	{
		ram = value;
	}
	public final int getCapacity()
	{
		return capacity;
	}
	public final void setCapacity(int value)
	{
		capacity = value;
	}
	public final String getColor()
	{
		return color;
	}
	public final void setColor(String value)
	{
		color = value;
	}
	public final double getPrice()
	{
		return price;
	}
	public final void setPrice(double value)
	{
		price = value;
	}
	public final String getName()
	{
		return name;
	}
	public final void setName(String value)
	{
		name = value;
	}
	public final String getOs()
	{
		return os;
	}
	public final void setOs(String value)
	{
		os = value;
	}

	public Device()
	{
	}
	protected Device(String brand, String name, String os, String size, String chip, int ram, int capacity, String color, double price)
	{
		setBrand(brand);
		setSize(size);
		setChip(chip);
		setRam(ram);
		setCapacity(capacity);
		setColor(color);
		setPrice(price);
		setName(name);
		setOs(os);
	}
	protected Device(String brand, String name, String size, String chip, int ram, int capacity, String color, double price)
	{
		setBrand(brand);
		setSize(size);
		setChip(chip);
		setRam(ram);
		setCapacity(capacity);
		setColor(color);
		setPrice(price);
		setName(name);
	}
	public abstract void Info();

	public static boolean opLessThan(Device phone1, Device phone2)
	{
		int point = 0;
		if (phone1.getRam() < phone2.getRam())
		{
			point++;
		}
		if (phone1.getCapacity() < phone2.getCapacity())
		{
			point++;
		}
		if (phone1.getPrice() > phone2.getPrice())
		{
			point++;
		}
		if (point >= 2)
		{
			return true;
		}
		return false;
	}
	public static boolean opGreaterThan(Device phone1, Device phone2)
	{
		int point = 0;
		if (phone1.getRam() > phone2.getRam())
		{
			point++;
		}
		if (phone1.getCapacity() > phone2.getCapacity())
		{
			point++;
		}
		if (phone1.getPrice() < phone2.getPrice())
		{
			point++;
		}
		if (point >= 2)
		{
			return true;
		}
		return false;
	}
	public String installApp(IAndroidCompatible app)
	{
		return "";
	}
	public String installApp(IAppleCompatible app)
	{
		return "";
	}

}
